hw/ip/aon_timer/rtl/aon_timer.sv

```
@@ -0,0 +1,435 @@
+    .d_i    (wkup_thold),
+    .q_o    (hw2reg.wkup_thold.d));
+
+  // wkup_count
+  prim_flop_2sync #(.Width (32)) wkup_count_rd_sync (
+    .clk_i  (clk_i),
+    .rst_ni (rst_ni),
+    .d_i    (wkup_count),
+    .q_o    (hw2reg.wkup_count.d));
+
+  // wdog_ctrl
+  prim_flop_2sync #(.Width (1)) wdog_ctrl_enable_rd_sync (
+    .clk_i  (clk_i),
+    .rst_ni (rst_ni),
+    .d_i    (wdog_enable),
+    .q_o    (hw2reg.wdog_ctrl.enable.d));
+
+  prim_flop_2sync #(.Width (1)) wdog_ctrl_pause_rd_sync (
+    .clk_i  (clk_i),
+    .rst_ni (rst_ni),
+    .d_i    (wdog_pause),
+    .q_o    (hw2reg.wdog_ctrl.pause_in_sleep.d));
+
+  // wdog_bark_thold
+  prim_flop_2sync #(.Width (32)) wdog_bark_thold_rd_sync (
+    .clk_i  (clk_i),
+    .rst_ni (rst_ni),
+    .d_i    (wdog_bark_thold),
+    .q_o    (hw2reg.wdog_bark_thold.d));
+
+  // wdog_bite_thold
+  prim_flop_2sync #(.Width (32)) wdog_bite_thold_rd_sync (
+    .clk_i  (clk_i),
+    .rst_ni (rst_ni),
+    .d_i    (wdog_bite_thold),
+    .q_o    (hw2reg.wdog_bite_thold.d));
+
+  // wdog_count
+  prim_flop_2sync #(.Width (32)) wdog_count_rd_sync (
+    .clk_i  (clk_i),
+    .rst_ni (rst_ni),
+    .d_i    (wdog_count),
+    .q_o    (hw2reg.wdog_count.d));
+
+  // wkup_cause
+  prim_flop_2sync #(.Width (1)) wkup_cause_rd_sync (
+    .clk_i  (clk_i),
+    .rst_ni (rst_ni),
+    .d_i    (aon_wkup_req_q),
+    .q_o    (hw2reg.wkup_cause.d));
+
+  //////////////////////////////
+  // Register Write Interface //
+  //////////////////////////////
+
+  // Register write data needs to be synchronized to make sure we capture sensible values.
+  // TODO This would probaby make more sense as a single fifo for all registers, but can't really
+  // achieve that at the moment with current register tooling.
+  // TODO An alternative improvement would be to fix the async fifo to allow depth < 3
+  // Note: these are fast clock to slow clock transfers
+
+  // wkup_ctrl
+  prim_fifo_async #(.Width (13), .Depth (3)) wkup_ctrl_wr_data_sync (
```

@eunchan 

yeah i definitely agree it would make sense to use a single FIFO for all registers.
Would you basically just sync the write enable / address / data through and do the decoding on the other end?

One thing to think about is how you would expect software to know the write is "done".  Because if we're doing this way, it kind of seems like software would get a tlul response that it's complete way before it actually is right? 

would it make sense for us to add some kind of 'done' for `hwext` type registers?

---

Once the fifo supports Depth==1 I can just use that. A fifo of Depth==1 per-register will be similar in area to one big fifo capable of buffering a write to every register (saving the width of the narrower registers plus the address etc should offset the extra control logic, which isn't much for the 1-entry fifo).

The "done" signal is an interesting one. We will indeed signal that the write is done a long time before the underlying register updates, but the alternative is to stall the bus and the cpu for 1000s of cycles on every register write. I have added a note to the documentation to warn the user of this, and to indicate that SW can read-back the register if it wants to know whether an update has propagated.

There's also a question of what to do if the user overflows the fifo by writing registers too quickly. At the moment those later writes are silently dropped. We could stall the bus in that case and allow the writes to propagate, or signal an error?

---

i personally think it's better to stall. Even though most of the time this should be wrapped under correct software behavior, I can definitely see scenarios where something does the wrong thing and this ends up a bit confusing when not using DIFs. 

Could you maybe check quickly with a few software folks on their end if they're okay with the "gate with read" scenario? Because I think if we require this, more likely than not they will not write a bunch of registers, and then check them one by one.  It seems more likely they will write one, read poll and then continue on.

---

Maybe the best thing is to stall on potential overflow but also raise some sort of interrupt? That way, production code can disable the interrupt, but we needn't worry about weird slow-downs when we're doing bring-up.

---

Adding @silvestrst and @gkelly for comment. My initial thinking was that software wouldn't particularly care when updates propagated through to the actual registers here since the configuration is pretty much fire and forget.

If we expect software to poll on completion of things then maybe the approach you highlighted in the pwrmgr with a single sync register would be better.

---

@tomroberts-lowrisc , i think Andrey's opinion was that software would poll on the completion of the final write (with the assumption that if this one completes, all previous ones have completed as well).  I think that's a reasonable assumption to make.

As for the thing to do here, if we go with the distributed FIFO approach, then i see less of a need for the `done` unless you think software will write to the same registers / field again and again? 

I think long term if we do what was discussed (ie, absorb a clock crossing FIFO into regfile itself, thus allowing certain registers to be in different domains), we can assume THAT FIFO will back-pressure if it becomes full.  So probably no need to get too fancy here and probably no need to switch schemes.  Can you double check to see if any of these crossings you expect software to write multiple times in quick succession?